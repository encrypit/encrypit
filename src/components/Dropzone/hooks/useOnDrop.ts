import { useCallback } from 'react';
import type { DropEvent, DropzoneOptions, FileRejection } from 'react-dropzone';
import { useNavigate } from 'react-router-dom';
import { API_URL } from 'src/config';
import { FILE } from 'src/constants';

type OnDrop = Required<DropzoneOptions>['onDrop'];

export function useOnDrop() {
  const navigate = useNavigate();

  const onDrop: OnDrop = useCallback(
    async (
      acceptedFiles: File[],
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      fileRejections: FileRejection[],
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      event: DropEvent
    ) => {
      if (fileRejections.length || !acceptedFiles.length) {
        return;
      }

      const formData = new FormData();
      formData.append(FILE, acceptedFiles[0]);

      const response = await fetch(`${API_URL}/api/files`, {
        body: formData,
        method: 'POST',
      });

      const uuid = await response.text();
      // eslint-disable-next-line no-console
      console.log(uuid);

      navigate('/share', { replace: true });
    },
    []
  );

  return onDrop;
}
