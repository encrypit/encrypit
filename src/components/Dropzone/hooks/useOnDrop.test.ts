import { renderHook } from '@testing-library/react';
import { mockFiles, store, wrapper } from 'test/helpers';

import { useOnDrop } from './useOnDrop';

const event = new Event('drop');

it('returns onDrop callback', () => {
  const { result } = renderHook(() => useOnDrop(), { wrapper });
  expect(result.current).toBeInstanceOf(Function);
});

describe('success', () => {
  it.each([1, 2])('sets %d files in store', async (count) => {
    const { result } = renderHook(() => useOnDrop(), { wrapper });
    const files = mockFiles(count);
    await result.current(files, [], event);
    expect(store.getState().file).toMatchSnapshot();
  });
});

describe('error', () => {
  it('does not upload if there is no file', async () => {
    const { result } = renderHook(() => useOnDrop(), { wrapper });
    await result.current([], [], event);
    expect(store.getState().file).toMatchInlineSnapshot(`
      {
        "files": [],
        "key": "",
      }
    `);
  });

  it('does not upload if there are file rejections', async () => {
    const { result } = renderHook(() => useOnDrop(), { wrapper });
    const files = mockFiles();
    const fileRejections = files.map((file) => ({
      errors: [{ code: 'too-many-files', message: 'Too many files' }],
      file,
    }));
    await result.current(files, fileRejections, event);
    expect(store.getState().file).toMatchInlineSnapshot(`
      {
        "files": [],
        "key": "",
      }
    `);
  });
});
