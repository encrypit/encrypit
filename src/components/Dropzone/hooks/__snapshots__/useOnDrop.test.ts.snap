// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`success sets 1 files in store 1`] = `
{
  "files": [
    {
      "data": "data:application/json;base64,eyJwaW5nIjp0cnVlfQ==",
      "id": "a7177311-3b7a-4593-942a-36b23c5afd30",
      "name": "ping.json",
      "size": 13,
      "type": "application/json",
    },
  ],
  "key": "",
  "password": "",
}
`;

exports[`success sets 2 files in store 1`] = `
{
  "files": [
    {
      "data": "data:application/json;base64,eyJwaW5nIjp0cnVlfQ==",
      "id": "a7177311-3b7a-4593-942a-36b23c5afd30",
      "name": "ping.json",
      "size": 13,
      "type": "application/json",
    },
    {
      "data": "data:application/json;base64,eyJwaW5nIjp0cnVlfQ==",
      "id": "a7177311-3b7a-4593-942a-36b23c5afd30",
      "name": "ping.json",
      "size": 13,
      "type": "application/json",
    },
  ],
  "key": "",
  "password": "",
}
`;
